File: Ch01/Ex1_09.scala
/**
  * Ex1_09.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * Define both a recursive and iterative process for addition
  * (Exercise 1.9)
  */

def inc(x: Int) = { x + 1 } /* TODO: figure out point-free */
def dec(x: Int) = { x - 1 } /* TODO: figure out point-free */

def rAdd(a: Int, b: Int): Int = {
  a match {
    case 0 => b
    case _ => inc(rAdd(dec(a), b))
  }
}
def iAdd(a: Int, b: Int): Int = {
  a match {
    case 0 => b
    case _ => iAdd(dec(a), inc(b))
  }
}

/**
  * execute with
  * scala -i Ex1_09.scala -e "println(rAdd(1,2) + iAdd(3,4))"
  * until I figure out a better way
  */

File: Ch01/ex1_09.exs
# SICP in Elixir: Exercise 1.9 - Define both a recursive and iterative process for addition

defmodule Ex1_09 do
  def i_add(0, b), do: b
  def i_add(a, b)  do
    i_add(dec(a), inc(b))
  end
  def r_add(0, b), do: b
  def r_add(a, b)  do
    inc(r_add(dec(a), b))
  end

# Private functions
  defp dec(x), do: x - 1
  defp inc(x), do: x + 1
end

IO.puts Ex1_09.i_add(5, 5)
IO.puts Ex1_09.r_add(5, 5)
